## STEP 1 Install Darknet


# Tutorial: Training a Yolov4 Object Detector with poker cards images



In this tutorial we will train an object detector using the Yolo v4 model. This model will run on Ultra96v2 board.

The model is pretrained on the COCO dataset. The framework used for training is Darknet.
Will run through the following steps:


*   Install the libraries (Darknet , etc.)
*   Clone the github repo and replace the repo training data with your data (from google drive or from own repo - which is faster)
*   Train the model on the new images


You can make a copy of this tutorial: File-> Save a copy in Drive

Note: the model training can be run with the repo images of 6 poker cards.If you choose to skip the customization part for your own images just to see how the training and the rest of the steps work.

In [ ]:
%%capture
!git clone https://github.com/AlexeyAB/darknet.git
%cd darknet
import re
!sed -i 's/OPENCV=0/OPENCV=1/' Makefile
!sed -i 's/GPU=0/GPU=1/' Makefile
!sed -i 's/CUDNN=0/CUDNN=1/' Makefile
# !sed -i 's/CUDNN_HALF=0/CUDNN_HALF=1/' Makefile
!make
!chmod +x ./darknet


In [ ]:
%tensorflow_version 1.x

TensorFlow 1.x selected.


In [ ]:
import tensorflow as tf
tf.__version__

'1.15.2'

In [ ]:
#utility function
def imShow(path):
  import cv2
  import matplotlib.pyplot as plt
  %matplotlib inline

  image = cv2.imread(path)
  height, width = image.shape[:2]
  resized_image = cv2.resize(image,(3*width, 3*height), interpolation = cv2.INTER_CUBIC)

  fig = plt.gcf()
  fig.set_size_inches(18, 10)
  plt.axis("off")
  #plt.rcParams['figure.figsize'] = [10, 5]
  plt.imshow(cv2.cvtColor(resized_image, cv2.COLOR_BGR2RGB))
  plt.show()
  
  


## Clone a reference/blueprint repo
This repo provide the flow from yolov4 model training to deployment.
In order to train on your own data, this repo can be used as a blueprint. 
Changes to some files need to be made according to own data, but it is nothing complicated.

In [1]:
!git clone https://github.com/brightcome/kids-math-study-ai-mate-on-ultra96v2
!ls


Cloning into 'kids-math-study-ai-mate-on-ultra96v2'...
remote: Enumerating objects: 109, done.
remote: Total 109 (delta 0), reused 0 (delta 0), pack-reused 109
Receiving objects: 100% (109/109), 224.83 MiB | 50.37 MiB/s, done.
Resolving deltas: 100% (5/5), done.
kids-math-study-ai-mate-on-ultra96v2  sample_data


### Now go to the file explorer on the left of the colab notebook and click on content then on the "kids-math-study-ai-mate-on-ultra96v2" folder.
Here you will see a folder and a few files.
The `train_images` folder contains both images AND the associated annotations in yolo format. 

(to annotate own images use https://github.com/tzutalin/labelImg and save annotations in yolo format)
The dk_files folder contains training related files and command.
* the 'obj.data' file contains paths for files used by Darknet and the number of classes to train the yolo model.

* the 'obj.names' file has the labels of the objects to be detected. the number of objects gives the number of classes in the 'obj.data' file.

* the train_list.txt contains the path to each image to be used in training. similarly, valid_list.txt contains the path to each image for testing. when not many images are available for training (a few hundred), use all images for training for better model performance. As a simple demo , I only have 40 images for training and validation. You should take pictures by your own and add more pictures into the folder.

The dk_model folder contains darknet yolov4 model file.
* the yolov4-leaky_best.weights.7z.001 are the COCO pretrained weights. Please unzip the package files. 

* the yolov4-leaky_obj.cfg are is the yolov4 configuration file that tells Darknet the model parameters and structure. the cfg file has been changed according to my demo requirement- 6 classes.

So you should change according to your data and classes, if needed.


# How to use your own images
In order to use your own annotated images, follow the steps below. If you want to train on the images already in the repo, skip this section and jump to the Train model section.

## Let's start with the images and annotations. I will use my images as example.
We copy the images on our google drive in a folder. My gdrive folder is named medmask_yolo. Then we mount the gdrive in colab and copy the images and annotations in the obj folder.

In [ ]:
!curl -L "https://app.roboflow.com/ds/yv8hShJnp0?key=fWKMkpQkwg" > roboflow.zip; unzip roboflow.zip; rm roboflow.zip

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   890  100   890    0     0   1163      0 --:--:-- --:--:-- --:--:--  1161
100 1088k  100 1088k    0     0  1138k      0 --:--:-- --:--:-- --:--:-- 1138k
Archive:  roboflow.zip
 extracting: README.roboflow.txt     
   creating: test/
 extracting: test/_darknet.labels    
 extracting: test/cam_image13_jpg.rf.b2c29d92454fe9bbe0a1f21a8224e150.jpg  
 extracting: test/cam_image13_jpg.rf.b2c29d92454fe9bbe0a1f21a8224e150.txt  
 extracting: test/cam_image17_jpg.rf.4120e59c7d0ea06aa6ae18f84fa86a9c.jpg  
 extracting: test/cam_image17_jpg.rf.4120e59c7d0ea06aa6ae18f84fa86a9c.txt  
 extracting: test/cam_image27_jpg.rf.ed2f40f84c347b595f05b24cb8aa176b.jpg  
 extracting: test/cam_image27_jpg.rf.ed2f40f84c347b595f05b24cb8aa176b.txt  
   creating: train/
 extracting: train/_darknet.labels   
 extracting: train/cam_image11_jpg.rf.467edeeb2678bf

In [14]:
#copy images to dk_files folder
#this can take a few minutes depending on dataset size. 
!pwd
#%cd ./dk_files/
!cp -rf './train_images/*.*' './dk_files/obj'

shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
pwd: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
shell-init: error retrieving current directory: getcwd: cannot access parent directories: No such file or directory
cp: cannot stat './train_images/*.*': No such file or directory


### generate image files list for training
The image convert package do NOT include list , so need generate.

In [ ]:
%cd /content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files/
!pwd
%cd ./train
!ls *.jpg > ../train_list.txt
%cd ../
%cd ./valid
!ls *.jpg > ../valid_list.txt
%cd ../

/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files/train
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files/valid
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files


extract the pre-trained yolov4 model weights

In [ ]:
#！sudo apt install p7zip-full
! pwd
%cd ./dk_model
!7z x yolov4-leaky_best.weights.7z.001

/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_model

7-Zip [64] 16.02 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-21
p7zip Version 16.02 (locale=en_US.UTF-8,Utf16=on,HugeFiles=on,64 bits,2 CPUs Intel(R) Xeon(R) CPU @ 2.20GHz (406F0),ASM,AES-NI)

Scanning the drive for archives:
  0M Scan         1 file, 24641535 bytes (24 MiB)

Extracting archive: yolov4-leaky_best.weights.7z.001
  0% 1 Open           --
Path = yolov4-leaky_best.weights.7z.001
Type = Split
Physical Size = 24641535
Volumes = 10
Total Physical Size = 234427775
----
Path = yolov4-leaky_best.weights.7z
Size = 234427775
--
Path = yolov4-leaky_best.weights.7z
Type = 7z
Physical Size = 234427775
Headers Size = 152
Method = LZMA:25
Solid = -
Blocks = 1

  0%      1%      2% - yolov4-leaky_best.weights         

## Change the labels in obj.names to our current labels
Or you can directly use or upload my obj.names 

In [ ]:
labels_path = '/content/yolotinyv3_medmask_demo/obj.names'
#make a list of your labels
labels = ['ace','jack','king','nine','queen','ten']


with open(labels_path, 'w') as f:

    f.write('\n'.join(labels))

#check that the labels file is correct
!cat /content/yolotinyv3_medmask_demo/obj.names

mask
no mask

## Change the number of classes in obj.data.
The paths are relative so no change there as long as the folder/file structure/names are not changed.

In [ ]:
import re
objdata = '/content/yolotinyv3_medmask_demo/obj.data'
with open(objdata) as f:
    s = f.read()

#the number of classes is equal to the number of labels
num_classes = len(labels)   
s = re.sub('classes = \d*','classes = ' + str(num_classes),s)

with open(objdata, 'w') as f:
  f.write(s)
!cat /content/yolotinyv3_medmask_demo/obj.data

classes= 2
train  = /content/yolotinyv3_medmask_demo/train.txt
valid  = /content/yolotinyv3_medmask_demo/test.txt
names = /content/yolotinyv3_medmask_demo/obj.names
backup = backup/

## Now we need to create a train.txt and test.txt file for our images from `obj` folder.
We will use a script that splits the images based on a specified percentage 
and writes them in train.txt and test.txt files.

In [ ]:
%cd ../yolotinyv3_medmask_demo/

/content/yolotinyv3_medmask_demo


In [ ]:
#in this case we use 6 percent as the number of images for testing.
#double clicking on the train or test text files will open the file in a new
#tab in colab and display the content.
!python3 folder2textYolo.py 5 /content/yolotinyv3_medmask_demo/obj

Your image file extension is: .jpg
Number of images: 762
Number of images used for training 724
Number of images used for testing 38


## Last thing we need to change is the yolov4-leaky_obj.cfg file.
There are several parameters of importance here. They control various aspects of the training process. Let's print the first few lines and have a look


In [ ]:
!head -n 24 /content/yolotinyv3_medmask_demo/yolov4-leaky_obj.cfg 


[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 4000
policy=steps
steps=3200,3600
scales=.1,.1



### Let's go over the parameters above:
* The batch parameter dictates the batch size. That one generally remains at 64.
* The subdivisions dictates how many images are loaded into memory. A smaller number translates to faster training. We will use 12. If a CUDA out of memory error is triggered, the subdivisions should be increased to i.e. 16,24,32 or some other number. (64 is max) Unless training with a resolution higher than 416, there should be no problem with 12.
* The width and height are by default at 416. Another resolution one can try is 320, or 608.
* The next parameter we care about is max_batches. This determines how long the training process is. Its value should be at least around 2000 for every class used. So for 3 classes, at least 6000.
* The steps are calculated as a function of max_batches. The first value is 0.8 * max_batches and the second value 0.9 * max_batches. In this case it is 8000 * 0.8 and 8000 * 0.9. 

For more details about Yolo parameters have a look here:
https://github.com/AlexeyAB/darknet#how-to-train-to-detect-your-custom-objects
There is a lot of material here, so make sure you scroll through if you have questions. Pretty much everything is well explained.




### Now that we see which parameters are of importance, let's modify them according to our dataset.

In [ ]:
# set the number of max_batches - min 2000 per class:
max_batch=4000
# calculate the 2 steps values:
step1 = 0.8 * max_batch
step2 = 0.9 * max_batch

# we also need to adjust the number of classes and a parameter called filter size 
# that are both is inside the model structure

# num_classes = len(labels)
num_filters = (num_classes + 5) * 3


cfg_file = '/content/yolotinyv3_medmask_demo/yolov4-leaky_obj.cfg'
# cfg_file = '/content/yolotinyv3_medmask_demo/yolov4-tiny.cfg'

with open(cfg_file) as f:
    s = f.read()
# (re.sub('[a-z]*@', 'ABC@', s))
s = re.sub('max_batches = \d*','max_batches = '+str(max_batch),s)
s = re.sub('steps=\d*,\d*','steps='+"{:.0f}".format(step1)+','+"{:.0f}".format(step2),s)
s = re.sub('classes=\d*','classes='+str(num_classes),s)
s = re.sub('pad=1\nfilters=\d*','pad=1\nfilters='+"{:.0f}".format(num_filters),s)
# pad=1\nfilters=\d\d
# s = re.sub('CUDNN=0','CUDNN=1',s)
# s = re.sub('OPENCV=0','OPENCV=1',s)

with open(cfg_file, 'w') as f:
  # s = re.sub('GPU=0','GPU=1',s)
  f.write(s)



Let's look again at the configuration 

In [ ]:
!head -n 24 /content/yolotinyv3_medmask_demo/yolov4-leaky_obj.cfg 


[net]
# Testing
#batch=1
#subdivisions=1
# Training
batch=64
subdivisions=16
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 4000
policy=steps
steps=3200,3600
scales=.1,.1



We also check if the number of classes and filters updated correctly

In [ ]:
!tail -n 64 /content/yolotinyv3_medmask_demo/yolov4-leaky_obj.cfg 


stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear



[yolo]
mask = 3,4,5
anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
classes=2
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1

[route]
layers = -4

[convolutional]
batch_normalize=1
filters=128
size=1
stride=1
pad=1
activation=leaky

[upsample]
stride=2

[route]
layers = -1, 8

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=21
activation=linear

[yolo]
mask = 1,2,3
anchors = 10,14,  23,27,  37,58,  81,82,  135,169,  344,319
classes=2
num=6
jitter=.3
ignore_thresh = .7
truth_thresh = 1
random=1


# Train the model

When you execute the following command, your model will start training.

You will have a log line per epoch. On each iteration you will see how your training is going.

In the `content/darknet/backup/` folder Darknet saves a few weights files:
* files that end in 1000,2000 etc are weights saved every 1000 batches
* best weights file are the weights that gave best results during training
* final weights file contains the final weights at the end of the training





## Start the model training

In [ ]:
!pwd
#%cd ../

/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files


In [ ]:
!source ./train_yolov4_colab.sh

Streaming output truncated to the last 5000 lines.
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.000001), count: 1, total_loss = 0.663017 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.627844), count: 21, total_loss = 14.708786 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 161 Avg (IOU: 0.586593), count: 36, total_loss = 16.691256 
 total_bbox = 382726, rewritten_bbox = 0.000000 % 

 (next mAP calculation at 1600 iterations) 
 Last accuracy mAP@0.5 = 25.09 %, best = 27.64 % 
 1554: 6.907785, 5.604225 avg loss, 0.000026 rate, 5.283232 seconds, 37296 images, 10.153673 hours left
Loaded: 0.000046 seconds
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 139 Avg (IOU: 0.000000), count: 1, total_loss = 0.000112 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 1.00, cls: 1.00) Region 150 Avg (IOU: 0.593167), count: 16, total_loss = 12.595292 
v3 (iou loss, Normalizer: (iou: 0.07, obj: 

## Check the model performance
The metrics run on the test images so they may not be fully representative, only indicative

In [ ]:
!pwd
%cd train 
!cp -rf ../valid/*.* ./

/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files
/content/Vitis-In-Depth-Tutorial/Machine_Learning/Design_Tutorials/07-yolov4-tutorial/dk_files/train


In [ ]:
!./darknet detector map /content/yolotinyv3_medmask_demo/obj.data /content/yolotinyv3_medmask_demo/yolov4-leaky_obj.cfg "/content/darknet/backup/yolov4-leaky_obj_best.weights" -points 0

## Copy best weights to google drive


In [ ]:
!cp /content/darknet/backup/yolov4-leaky_obj_best.weights  '/content/drive/My Drive/'